syntax = "proto2";

package RTDBM;

import "common.proto";

// Коды используемых элементарных типов данных eXtremeDB
// Оригинал в файле xdb_common.hpp
enum DbType {
  DB_TYPE_UNDEF     = 0;
  DB_TYPE_LOGICAL   = 1;
  DB_TYPE_INT8      = 2;
  DB_TYPE_UINT8     = 3;
  DB_TYPE_INT16     = 4;
  DB_TYPE_UINT16    = 5;
  DB_TYPE_INT32     = 6;
  DB_TYPE_UINT32    = 7;
  DB_TYPE_INT64     = 8;
  DB_TYPE_UINT64    = 9;
  DB_TYPE_FLOAT     = 10;
  DB_TYPE_DOUBLE    = 11;
  DB_TYPE_BYTES     = 12; // переменная длина строки
  DB_TYPE_BYTES4    = 13;
  DB_TYPE_BYTES8    = 14;
  DB_TYPE_BYTES12   = 15;
  DB_TYPE_BYTES16   = 16;
  DB_TYPE_BYTES20   = 17;
  DB_TYPE_BYTES32   = 18;
  DB_TYPE_BYTES48   = 19;
  DB_TYPE_BYTES64   = 20;
  DB_TYPE_BYTES80   = 21;
  DB_TYPE_BYTES128  = 22;
  DB_TYPE_BYTES256  = 23;
  DB_TYPE_ABSTIME   = 24;
}

// Качество атрибута в БДРВ, оригинал находится в xdb_common.hpp
// NB: синхронизовать изменения с оригиналом
enum Quality {
  ATTR_OK        = 0; /* no known error on data */
  ATTR_SUSPECT   = 1; /* depends on a suspect, error or disabled value */
  ATTR_ERROR     = 2; /* calc. engine got math error */
  ATTR_DISABLED  = 3; /* calc. engine operation indicator disabled */
  ATTR_NOT_FOUND = 4; /* attribute is not found in database */
}

// Тип выбираемой истории
// NB: синхронизировать со схемой БДРВ (rtap_db.mco, HistoryType)
enum HistoryType {
  PER_NONE       = 0;
  PER_1_MINUTE   = 1;
  PER_5_MINUTES  = 2;
  PER_HOUR       = 3;
  PER_DAY        = 4;
  PER_MONTH      = 5;
}

message ValueUpdate {
  required string   tag     = 1;
  required DbType   type    = 2 [default=DB_TYPE_UNDEF];
  required Quality  quality = 3 [default=ATTR_OK];
  optional int64    i64_value   = 4;
  optional int32    i32_value   = 5;
  optional float    f_value     = 6;
  optional double   g_value     = 7;
  optional string   s_value     = 8;
  optional bool     b_value     = 9;
}

// Чтение одного или более значения из БДРВ
// #SIG_D_MSG_READ_MULTI
message ReadMulti {
  repeated ValueUpdate  update_item = 1;
}

// Запись одного или более значений в БДРВ
// #SIG_D_MSG_WRITE_MULTI
message WriteMulti {
  repeated ValueUpdate  update_item = 1;
}

// Управление подписками
// #SIG_D_MSG_GRPSBS_CTRL
message SubscriptionControl {
  // Название группы
  required string   name    = 1;
  // Код управления {ENABLE|SUSPEND}
  required int32    ctrl    = 2;
}

// Обновление от сервера при изменении одного из атрибутов Группы Подписки
// #SIG_D_MSG_GRPSBS
message SubscriptionEvent {
  repeated ValueUpdate  update_item = 1;
}

// Элементарное значение предыстории
message HistoryItem {
  // Достоверность отсчёта
  required gof_t_AcqInfoValid valid = 1 [default=GOF_D_ACQ_INVALID];
  // Время формирования
  required int32  frame = 2 [default=0];
  // Значение отсчёта
  required double value = 3 [default=0];
}

// Запрос получения истории указанного типа для указанного тега 
// #SIG_D_MSG_REQ_HISTORY
message HistoryRequest {
  // Запрашиваемая история для указанного Тега 
  required string   tag         = 1;
  // Время начала интересующего интервала истории
  required int32    time_start  = 2;
  // Количество запрашиваемых семплов
  required int32    num_required_samples = 3;
  // Тип выбираемой истории (минутная, 5-ти минутная, часовая, ...)
  required HistoryType history_type = 4;
  // Массив значений, заполняется при ответе
  repeated HistoryItem item = 5;
}

